From 16111da2f4b28945a85716f7137ac59f6c559d97 Mon Sep 17 00:00:00 2001
From: "choongryeol.lee" <choongryeol.lee@lge.com>
Date: Fri, 25 Oct 2013 17:44:22 -0700
Subject: [PATCH 032/507] power: max17048: fix having old soc when the system
 is resumed

Since it takes some time to restart the battery monitor work from
the system resume, there is a brief period when the battery level
has a value before suspend. To fix this, wait until the battery
level is updated before returning the capacity if the battery
monitor work is not started yet.

Change-Id: Ibbebcd653dd476d36a24c6b89ed8e4c0a00391be
Signed-off-by: choongryeol.lee <choongryeol.lee@lge.com>
Signed-off-by: Simarpreet Singh <simar@linux.com>
---
 drivers/power/max17048_battery.c |   22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/drivers/power/max17048_battery.c b/drivers/power/max17048_battery.c
index 30c13aa..7c5e9fa 100644
--- a/drivers/power/max17048_battery.c
+++ b/drivers/power/max17048_battery.c
@@ -87,6 +87,7 @@ struct max17048_chip {
 };
 
 static struct max17048_chip *ref;
+struct completion monitor_work_done;
 static int max17048_get_prop_current(struct max17048_chip *chip);
 static int max17048_get_prop_temp(struct max17048_chip *chip);
 static int max17048_clear_interrupt(struct max17048_chip *chip);
@@ -338,6 +339,7 @@ static void max17048_work(struct work_struct *work)
 
 	max17048_get_vcell(chip);
 	max17048_get_soc(chip);
+	complete_all(&monitor_work_done);
 
 	if (chip->capacity_level != chip->lasttime_capacity_level ||
 			chip->capacity_level == 0) {
@@ -659,6 +661,21 @@ static int max17048_get_prop_current(struct max17048_chip *chip)
 	return chip->batt_current;
 }
 
+static int max17048_get_prop_capacity(struct max17048_chip *chip)
+{
+	int ret;
+
+	if (chip->capacity_level == -EINVAL)
+		return -EINVAL;
+
+	ret = wait_for_completion_timeout(&monitor_work_done,
+					msecs_to_jiffies(500));
+	if (!ret)
+		pr_err("%s: timeout monitor work done\n", __func__);
+
+	return chip->capacity_level;
+}
+
 static enum power_supply_property max17048_battery_props[] = {
 	POWER_SUPPLY_PROP_STATUS,
 	POWER_SUPPLY_PROP_HEALTH,
@@ -706,7 +723,7 @@ static int max17048_get_property(struct power_supply *psy,
 		val->intval = max17048_get_prop_temp(chip);
 		break;
 	case POWER_SUPPLY_PROP_CAPACITY:
-		val->intval = chip->capacity_level;
+		val->intval = max17048_get_prop_capacity(chip);
 		break;
 	case POWER_SUPPLY_PROP_CURRENT_NOW:
 		val->intval = max17048_get_prop_current(chip);
@@ -805,6 +822,7 @@ static int max17048_pm_notifier(struct notifier_block *notifier,
 		cancel_delayed_work_sync(&chip->monitor_work);
 		break;
 	case PM_POST_SUSPEND:
+		INIT_COMPLETION(monitor_work_done);
 		schedule_delayed_work(&chip->monitor_work,
 					msecs_to_jiffies(200));
 		max17048_set_alsc_alert(chip, true);
@@ -865,6 +883,8 @@ static int max17048_probe(struct i2c_client *client,
 		goto error;
 	}
 
+	init_completion(&monitor_work_done);
+	chip->capacity_level = -EINVAL;
 	ref = chip;
 	chip->batt_psy.name = "battery";
 	chip->batt_psy.type = POWER_SUPPLY_TYPE_BATTERY;
-- 
1.7.9.5

